src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -72,39 +73,68 @@ class TaskReplaceActor(
+      // the v2 run spec. We then schedule 2 more v2 instances. In the future we probably want to bind instances to a
+      // certain run spec. Until then we have to update the run spec in a [[TaskLauncherActor]]
+      await(launchQueue.sync(runSpec))
 
-    // start new instances, if possible
-    launchInstances()
+      // kill old instances to free some capacity
+      for (_ <- 0 until ignitionStrategy.nrToKillImmediately) killNextOldInstance()
 
-    // reset the launch queue delay
-    logger.info("Resetting the backoff delay before restarting the runSpec")
-    launchQueue.resetDelay(runSpec)
+      // start new instances, if possible
+      val launched = await(launchInstances())
 
-    // it might be possible, that we come here, but nothing is left to do
-    checkFinished()
+      // reset the launch queue delay
+      logger.info("Resetting the backoff delay before restarting the runSpec")
+      launchQueue.resetDelay(runSpec)
+
+      // it might be possible, that we come here, but nothing is left to do
+      checkFinished()
+
+      Done
+    }.pipeTo(self)
   }
 
   override def postStop(): Unit = {
     eventBus.unsubscribe(self)
     super.postStop()
   }
 
-  override def receive: Receive = readinessBehavior orElse replaceBehavior
+  override def receive: Receive = initializing
+
+  private def initializing: Receive = {
+    case Done =>
+      context.become(initialized)
+      unstashAll()
+
+    case Status.Failure(cause) =>
+      // escalate this failure
+      throw new IllegalStateException("while loading tasks", cause)
+
+    case stashMe: AnyRef =>
+      stash()
+  }
+
+  private def initialized: Receive = readinessBehavior orElse replaceBehavior
 
   def replaceBehavior: Receive = {
     // New instance failed to start, restart it
-    case InstanceChanged(id, `version`, `pathId`, condition, instance) if !oldInstanceIds(id) && considerTerminal(condition) =>
-      logger.error(s"New instance $id failed on agent ${instance.agentInfo.agentId} during app $pathId restart")
+    case InstanceChanged(id, `version`, `pathId`, condition, Instance(_, Some(agentInfo), _, _, _, _, reservation)) if !oldInstanceIds(id) && considerTerminal(condition) =>
+      logger.warn(s"New instance $id is terminal on agent ${agentInfo.agentId} during app $pathId restart: $condition reservation: $reservation")
       instanceTerminated(id)
       instancesStarted -= 1
-      launchInstances()
+      launchInstances().pipeTo(self)
 
     // Old instance successfully killed
     case InstanceChanged(id, _, `pathId`, condition, _) if oldInstanceIds(id) && considerTerminal(condition) =>
+      logger.info(s"Instance $id became $condition. Launching more instances.")
       oldInstanceIds -= id
-      launchInstances().foreach(_ => checkFinished())
+      instanceTerminated(id)
+      launchInstances().pipeTo(self).foreach(_ => checkFinished())
```

Comment:
This is a leak of actor internals. `.foreach(_ => checkFinished())` will be executed in the future context, not the actor one.

Reply:
Good catch!